File: services/priceUtils.ts
/**
 * Robust price normalization and formatting utilities
 * Following best practices for handling international price formats
 */

export interface PriceData {
  /** Normalized numeric value (no thousands separators, decimal point format) */
  numericValue: number;
  /** Original raw price string */
  rawValue: string;
  /** Detected locale format */
  detectedLocale: 'US' | 'EU' | 'UNKNOWN';
  /** Currency code if detected */
  currency?: string;
}

/**
 * Step 1: Strip all non-numeric characters except decimal separators
 */
const stripNonNumeric = (priceStr: string): string => {
  if (!priceStr || typeof priceStr !== 'string') return '0';
  
  // Remove currency symbols, spaces, letters, but keep dots and commas
  return priceStr.replace(/[^\d.,]/g, '');
};

/**
 * Step 2: Detect locale format based on separator patterns
 */
const detectLocaleFormat = (cleanPrice: string): 'US' | 'EU' | 'UNKNOWN' => {
  if (!cleanPrice) return 'UNKNOWN';
  
  const dotIndex = cleanPrice.lastIndexOf('.');
  const commaIndex = cleanPrice.lastIndexOf(',');
  
  // If both separators exist, the last one is usually decimal
  if (dotIndex > -1 && commaIndex > -1) {
    return dotIndex > commaIndex ? 'US' : 'EU';
  }
  
  // If only one separator exists
  if (dotIndex > -1) {
    const afterDot = cleanPrice.substring(dotIndex + 1);
    // If 3+ digits after dot, it's likely thousands separator (EU style)
    // If 1-2 digits, it's likely decimal separator (US style)
    return afterDot.length > 2 ? 'EU' : 'US';
  }
  
  if (commaIndex > -1) {
    const afterComma = cleanPrice.substring(commaIndex + 1);
    // If 3+ digits after comma, it's likely thousands separator (US style)
    // If 1-2 digits, it's likely decimal separator (EU style)
    return afterComma.length > 2 ? 'US' : 'EU';
  }
  
  return 'UNKNOWN';
};

/**
 * Step 3: Normalize to standard format (decimal point, no thousands)
 */
const normalizeToStandard = (cleanPrice: string, locale: 'US' | 'EU' | 'UNKNOWN'): string => {
  if (!cleanPrice) return '0';
  
  switch (locale) {
    case 'US':
      // US: 1,234.56 -> 1234.56
      return cleanPrice.replace(/,/g, '');
      
    case 'EU':
      // EU: 1.234,56 -> 1234.56
      const lastCommaIndex = cleanPrice.lastIndexOf(',');
      if (lastCommaIndex > -1) {
        const beforeComma = cleanPrice.substring(0, lastCommaIndex).replace(/\./g, '');
        const afterComma = cleanPrice.substring(lastCommaIndex + 1);
        return `${beforeComma}.${afterComma}`;
      }
      // No comma, just remove dots used as thousands separators
      return cleanPrice.replace(/\./g, '');
      
    case 'UNKNOWN':
    default:
      // Best effort: assume single separator is decimal
      const dotIndex = cleanPrice.lastIndexOf('.');
      const commaIndex = cleanPrice.lastIndexOf(',');
      
      if (dotIndex > commaIndex && dotIndex > -1) {
        return cleanPrice.replace(/,/g, '');
      } else if (commaIndex > -1) {
        const beforeComma = cleanPrice.substring(0, commaIndex).replace(/\./g, '');
        const afterComma = cleanPrice.substring(commaIndex + 1);
        return `${beforeComma}.${afterComma}`;
      }
      
      return cleanPrice.replace(/[.,]/g, '');
  }
};

/**
 * Step 4: Convert to safe numeric type
 */
const toNumeric = (normalizedStr: string): number => {
  if (!normalizedStr) return 0;
  
  const num = parseFloat(normalizedStr);
  return isNaN(num) ? 0 : num;
};

/**
 * Extract currency from original price string
 */
const extractCurrency = (originalStr: string): string | undefined => {
  if (!originalStr) return undefined;
  
  const currencyPatterns = [
    /USD/i,
    /EUR?/i,
    /IQD/i,
    /د\.ع/,
    /\$/,
    /€/,
    /£/,
    /¥/
  ];
  
  for (const pattern of currencyPatterns) {
    const match = originalStr.match(pattern);
    if (match) {
      const found = match[0];
      if (found === 'د.ع') return 'IQD';
      if (found === '$') return 'USD';
      if (found === '€') return 'EUR';
      if (found === '£') return 'GBP';
      if (found === '¥') return 'JPY';
      return found.toUpperCase();
    }
  }
  
  return undefined;
};

/**
 * Main price parsing function
 * Converts any price string to normalized numeric value
 */
export const parsePrice = (priceInput: string | number | null | undefined): PriceData => {
  // Handle numeric input
  if (typeof priceInput === 'number') {
    return {
      numericValue: isNaN(priceInput) ? 0 : priceInput,
      rawValue: String(priceInput),
      detectedLocale: 'UNKNOWN'
    };
  }
  
  // Handle null/undefined
  if (!priceInput) {
    return {
      numericValue: 0,
      rawValue: '0',
      detectedLocale: 'UNKNOWN'
    };
  }
  
  const rawValue = String(priceInput);
  
  // Step 1: Strip non-numeric
  const cleaned = stripNonNumeric(rawValue);
  
  // Step 2: Detect locale
  const locale = detectLocaleFormat(cleaned);
  
  // Step 3: Normalize
  const normalized = normalizeToStandard(cleaned, locale);
  
  // Step 4: Convert to numeric
  const numericValue = toNumeric(normalized);
  
  // Extract currency
  const currency = extractCurrency(rawValue);
  
  return {
    numericValue,
    rawValue,
    detectedLocale: locale,
    currency
  };
};

/**
 * Format price for display with proper locale formatting
 */
export const formatPrice = (
  numericValue: number, 
  currency: string = 'IQD',
  locale: 'US' | 'EU' = 'US'
): string => {
  if (isNaN(numericValue) || numericValue < 0) return '0';
  
  // Round to 2 decimal places for display
  const rounded = Math.round(numericValue * 100) / 100;
  
  if (locale === 'EU') {
    // EU format: 1.234,56
    const parts = rounded.toFixed(2).split('.');
    const wholePart = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, '.');
    const decimalPart = parts[1];
    
    // Only show decimals if not .00
    if (decimalPart === '00') {
      return wholePart;
    }
    return `${wholePart},${decimalPart}`;
  } else {
    // US format: 1,234.56
    const parts = rounded.toFixed(2).split('.');
    const wholePart = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    const decimalPart = parts[1];
    
    // Only show decimals if not .00
    if (decimalPart === '00') {
      return wholePart;
    }
    return `${wholePart}.${decimalPart}`;
  }
};

/**
 * Calculate discount percentage between two prices
 */
export const calculateDiscount = (originalPrice: number, salePrice: number): number => {
  if (!originalPrice || !salePrice || originalPrice <= salePrice) return 0;
  
  return Math.round(((originalPrice - salePrice) / originalPrice) * 100);
};

/**
 * Calculate savings amount
 */
export const calculateSavings = (originalPrice: number, salePrice: number): number => {
  if (!originalPrice || !salePrice || originalPrice <= salePrice) return 0;
  
  return originalPrice - salePrice;
};

/**
 * Batch process multiple prices for consistent normalization
 */
export const batchParseprices = (prices: (string | number | null | undefined)[]): PriceData[] => {
  return prices.map(price => parsePrice(price));
};

/**
 * Validate if a price string can be parsed reliably
 */
export const isValidPrice = (priceInput: string | number | null | undefined): boolean => {
  const parsed = parsePrice(priceInput);
  return parsed.numericValue > 0 && parsed.detectedLocale !== 'UNKNOWN';
};

File: backend/price_utils.py
"""
Robust price normalization and formatting utilities for backend
Following best practices for handling international price formats
"""

import re
from typing import Union, Optional, Dict, Any


def strip_non_numeric(price_str: str) -> str:
    """Step 1: Strip all non-numeric characters except decimal separators"""
    if not price_str or not isinstance(price_str, str):
        return '0'
    
    # Remove currency symbols, spaces, letters, but keep dots and commas
    return re.sub(r'[^\d.,]', '', price_str)


def detect_locale_format(clean_price: str) -> str:
    """Step 2: Detect locale format based on separator patterns"""
    if not clean_price:
        return 'UNKNOWN'
    
    dot_index = clean_price.rfind('.')
    comma_index = clean_price.rfind(',')
    
    # If both separators exist, the last one is usually decimal
    if dot_index > -1 and comma_index > -1:
        return 'US' if dot_index > comma_index else 'EU'
    
    # If only one separator exists
    if dot_index > -1:
        after_dot = clean_price[dot_index + 1:]
        # If 3+ digits after dot, it's likely thousands separator (EU style)
        # If 1-2 digits, it's likely decimal separator (US style)
        return 'EU' if len(after_dot) > 2 else 'US'
    
    if comma_index > -1:
        after_comma = clean_price[comma_index + 1:]
        # If 3+ digits after comma, it's likely thousands separator (US style)
        # If 1-2 digits, it's likely decimal separator (EU style)
        return 'US' if len(after_comma) > 2 else 'EU'
    
    return 'UNKNOWN'


def normalize_to_standard(clean_price: str, locale: str) -> str:
    """Step 3: Normalize to standard format (decimal point, no thousands)"""
    if not clean_price:
        return '0'
    
    if locale == 'US':
        # US: 1,234.56 -> 1234.56
        return clean_price.replace(',', '')
    
    elif locale == 'EU':
        # EU: 1.234,56 -> 1234.56
        last_comma_index = clean_price.rfind(',')
        if last_comma_index > -1:
            before_comma = clean_price[:last_comma_index].replace('.', '')
            after_comma = clean_price[last_comma_index + 1:]
            return f"{before_comma}.{after_comma}"
        # No comma, just remove dots used as thousands separators
        return clean_price.replace('.', '')
    
    else:  # UNKNOWN
        # Best effort: assume single separator is decimal
        dot_index = clean_price.rfind('.')
        comma_index = clean_price.rfind(',')
        
        if dot_index > comma_index and dot_index > -1:
            return clean_price.replace(',', '')
        elif comma_index > -1:
            before_comma = clean_price[:comma_index].replace('.', '')
            after_comma = clean_price[comma_index + 1:]
            return f"{before_comma}.{after_comma}"
        
        return clean_price.replace(',', '').replace('.', '')


def to_numeric(normalized_str: str) -> float:
    """Step 4: Convert to safe numeric type"""
    if not normalized_str:
        return 0.0
    
    try:
        num = float(normalized_str)
        return num if not (num != num) else 0.0  # Check for NaN
    except (ValueError, TypeError):
        return 0.0


def extract_currency(original_str: str) -> Optional[str]:
    """Extract currency from original price string"""
    if not original_str:
        return None
    
    currency_patterns = [
        (r'USD', 'USD'),
        (r'EUR?', 'EUR'),
        (r'IQD', 'IQD'),
        (r'د\.ع', 'IQD'),
        (r'\$', 'USD'),
        (r'€', 'EUR'),
        (r'£', 'GBP'),
        (r'¥', 'JPY')
    ]
    
    for pattern, currency in currency_patterns:
        if re.search(pattern, original_str, re.IGNORECASE):
            return currency
    
    return None


def parse_price(price_input: Union[str, int, float, None]) -> Dict[str, Any]:
    """
    Main price parsing function
    Converts any price string to normalized numeric value
    """
    # Handle numeric input
    if isinstance(price_input, (int, float)):
        numeric_value = float(price_input) if not (price_input != price_input) else 0.0
        return {
            'numeric_value': numeric_value,
            'raw_value': str(price_input),
            'detected_locale': 'UNKNOWN',
            'currency': None
        }
    
    # Handle null/None
    if not price_input:
        return {
            'numeric_value': 0.0,
            'raw_value': '0',
            'detected_locale': 'UNKNOWN',
            'currency': None
        }
    
    raw_value = str(price_input)
    
    # Step 1: Strip non-numeric
    cleaned = strip_non_numeric(raw_value)
    
    # Step 2: Detect locale
    locale = detect_locale_format(cleaned)
    
    # Step 3: Normalize
    normalized = normalize_to_standard(cleaned, locale)
    
    # Step 4: Convert to numeric
    numeric_value = to_numeric(normalized)
    
    # Extract currency
    currency = extract_currency(raw_value)
    
    return {
        'numeric_value': numeric_value,
        'raw_value': raw_value,
        'detected_locale': locale,
        'currency': currency
    }


def calculate_discount(original_price: float, sale_price: float) -> int:
    """Calculate discount percentage between two prices"""
    if not original_price or not sale_price or original_price <= sale_price:
        return 0
    
    return round(((original_price - sale_price) / original_price) * 100)


def calculate_savings(original_price: float, sale_price: float) -> float:
    """Calculate savings amount"""
    if not original_price or not sale_price or original_price <= sale_price:
        return 0.0
    
    return original_price - sale_price